raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -311,8 +311,18 @@ private void updateListenersProgress(long highWatermark) {
     private void updateListenersProgress(List<ListenerContext> listenerContexts, long highWatermark) {
         for (ListenerContext listenerContext : listenerContexts) {
             listenerContext.nextExpectedOffset().ifPresent(nextExpectedOffset -> {
-                if (nextExpectedOffset < log.startOffset()) {
-                    listenerContext.fireHandleSnapshot(log.startOffset());
+                if (nextExpectedOffset < log.startOffset() && nextExpectedOffset < highWatermark) {
+                    SnapshotReader<T> snapshot = earliestSnapshot().orElseThrow(() -> {
```

Comment:
Why would we use the earliest snapshot and not the latest?

Reply:
No good reason. I'll change it to use the latest.